### ISSUE TITLE
@overload doesn't work with @asynccontextmanager

### ISSUE THREAD

**Issue by alenzo-arch** [python/mypy#17340]
**Bug Report**

When trying to annotate overloads for an asynccontextmanager, I think mypy incorrectly reports an invalid return type.

**To Reproduce**

```python
from contextlib import asynccontextmanager, contextmanager
from typing import AsyncIterator, Iterator, overload

@overload
@asynccontextmanager
async def test_async() -> AsyncIterator[int]: ...


@overload
@asynccontextmanager
async def test_async(val: int = ...) -> AsyncIterator[int]: ...


@asynccontextmanager
async def test_async(val: int = 1) -> AsyncIterator[int]:
    yield val


@overload
@contextmanager
def test_sync() -> Iterator[int]: ...


@overload
@contextmanager
def test_sync(val: int = ...) -> Iterator[int]: ...


@contextmanager
def test_sync(val: int = 1) -> Iterator[int]:
    yield val
```

This gives an error of "Argument 1 to "asynccontextmanager" has incompatible type "Callable[[int], Coroutine[Any, Any, AsyncIterator[int]]]"; expected "Callable[[int], AsyncIterator[Never]]"

Where the synchronous version checks fine.

**Expected Behavior**

I expect the async version to work the same as the sync version

**Actual Behavior**

Mypy reports and error

```bash
mypy test.py
test.py:5: error: Argument 1 to "asynccontextmanager" has incompatible type "Callable[[], Coroutine[Any, Any, AsyncIterator[int]]]"; expected "Callable[[], AsyncIterator[Never]]"  [arg-type]
test.py:10: error: Argument 1 to "asynccontextmanager" has incompatible type "Callable[[int], Coroutine[Any, Any, AsyncIterator[int]]]"; expected "Callable[[int], AsyncIterator[Never]]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```
**Your Environment**

- Mypy version used: 1.10.0
- Mypy command-line flags: `mypy test.py`
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: Python 3.11.4

**Comment by sterliakov** [comment]
Funny enough, the following typechecks (note how `async` disappears in `overload` stubs). This seems to be the same kind of problem as functions/methods and decorators: too much of "indirectly applied" stuff confuses the checker. In your original snippet, the function is "async twice": `mypy` thinks that it returns a `Coroutine` which, when awaited, produces an `AsyncIterator`. But this only happens when applying a decorator to overloaded definition (see the full playground, `reveal_type` produces a right type indeed).

```python
from contextlib import asynccontextmanager, contextmanager
from typing import AsyncIterator, Iterator, overload

@overload
@asynccontextmanager
def test_async() -> AsyncIterator[int]: ...


@overload
@asynccontextmanager
def test_async(val: int = ...) -> AsyncIterator[int]: ...


@asynccontextmanager
async def test_async(val: int = 1) -> AsyncIterator[int]:
    yield val
```

(I'm posting this more like a temporary workaround, this definitely looks like a typechecker bug to me, and `pyright` has no objections to your snippet)

**Comment by alenzo-arch** [comment]
@sterliakov Thank for for addressing this as well as providing a temporary work around
